ooni/reporter.py

```diff
@@ -298,8 +299,13 @@ def createReport(self):
             self.agent = SOCKS5Agent(reactor, proxyEndpoint=proxyEndpoint)
 
         elif self.collectorAddress.startswith('https://'):
-            # XXX add support for securely reporting to HTTPS collectors.
-            log.err("HTTPS based collectors are currently not supported.")
+            # not sure if there's something else it needs.  Seems to work.
+            # Very difficult to get it to work with self-signed certs.
+            self.agent = Agent(reactor)
```

Comment:
I will look more into this as I have not yet checked out twisted 15.

Some time ago I started writing this piece of software: https://github.com/hellais/sslpin to be used perhaps also inside of ooniprobe (though the ambitions could be perhaps even a bit larger).

It implements self authenticating urls where the fingerprint of the cert are inside of the URI itself. You can see an example of how it works here: https://github.com/hellais/sslpin/blob/master/examples/get_url.py.

I don't think it will work with recent versions of twisted, but it's perhaps still some useful code to look at to see how to properly implement cert validation in twisted.